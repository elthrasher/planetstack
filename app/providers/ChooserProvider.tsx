import React, { createContext, ReactNode, useState } from 'react';

export enum ChooserMode {
  NONE,
  BACKGROUND,
  ICON,
  INSTRUCTIONS,
}

export interface ChooserContextModel {
  chooserMode: ChooserMode;
  setChooserMode: (chooserMode: ChooserMode) => void;
}

export const ChooserContext = createContext<ChooserContextModel>({
  chooserMode: ChooserMode.NONE,
  setChooserMode: /* istanbul ignore next */ () => null,
});

export const ChooserProvider = ({ children }: { children: ReactNode }): JSX.Element => {
  const [chooserMode, setChooserMode] = useState<ChooserMode>(ChooserMode.NONE);

  return <ChooserContext.Provider value={{ chooserMode, setChooserMode }}>{children}</ChooserContext.Provider>;
};
