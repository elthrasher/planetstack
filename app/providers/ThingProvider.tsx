import React, { createContext, ReactNode, useState } from 'react';
import useWebSocket from 'react-use-websocket';

const socketUrl = 'wss://jdnif5uxok.execute-api.us-east-1.amazonaws.com/dev';

export interface Thing {
  id: string;
  img: string;
  x: number;
  y: number;
}

export interface ThingContextModel {
  things: Thing[];
  addThing: (thing: Thing) => void;
  removeThing: (thing: Thing) => void;
}

export const ThingContext = createContext<ThingContextModel>({
  things: [],
  addThing: () => null,
  removeThing: () => null,
});

export const ThingProvider = ({ children }: { children: ReactNode }) => {
  const [things, setThings] = useState<Thing[]>([]);
  const addThing = (thing: Thing) => setThings([...things, thing]);
  const removeThing = (thing: Thing) => setThings(things.filter((t) => t !== thing));
  const { sendMessage, sendJsonMessage, lastMessage, lastJsonMessage, readyState, getWebSocket } = useWebSocket(
    socketUrl,
    {
      onError: (e) => console.error(e),
      onOpen: () => console.log(),
      //Will attempt to reconnect on all close events, such as server shutting down
      shouldReconnect: () => true,
    },
  );
  console.log(readyState);
  return <ThingContext.Provider value={{ things, addThing, removeThing }}>{children}</ThingContext.Provider>;
};
