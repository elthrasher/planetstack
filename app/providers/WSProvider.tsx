import React, { createContext, ReactNode, useState } from 'react';
import useWebSocket, { SendMessage } from 'react-use-websocket';

import { GameStateModel } from '../../types/gamestate';
import { MessageAction } from '../../types/MessageAction';

/* istanbul ignore next */
const getSocketUrl = async () => {
  const response = await fetch('./config.json');
  const config = await response.json();
  return `${config.socketUrl}/${config.stageName}`;
};

const defaultGs: GameStateModel = { bg: 0, icons: {} };

export interface WSContextModel {
  gs: GameStateModel;
  sendMessage: SendMessage;
}

export const WSContext = createContext<WSContextModel>({
  gs: defaultGs,
  sendMessage: () => null,
});

export const WSProvider = ({ children }: { children: ReactNode }): JSX.Element => {
  const [gs, setGs] = useState(defaultGs);
  /* istanbul ignore next */
  const { sendMessage } = useWebSocket(getSocketUrl, {
    onOpen: () => sendMessage(JSON.stringify({ action: MessageAction.GET_STATE })),
    onError: (e) => console.error(e),
    onMessage: (message: MessageEvent<string>) => setGs(JSON.parse(message.data)),
    share: true,
    shouldReconnect: () => true,
  });
  return <WSContext.Provider value={{ gs, sendMessage }}>{children}</WSContext.Provider>;
};
