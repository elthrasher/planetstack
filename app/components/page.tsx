import React, { useContext } from 'react';

import { backgrounds } from '../../types/data/backgrounds';
import { WSContext } from '../providers/WSProvider';
import { Chooser } from './chooser';
import { DraggableIcon } from './draggable-icon';
import { Header } from './header';

export const Page = (): JSX.Element => {
  const { gs } = useContext(WSContext);
  return (
    <div
      style={{
        backgroundImage: `url('${backgrounds[gs.bg].path}')`,
        backgroundRepeat: 'no-repeat',
        backgroundSize: 'cover',
        display: 'block',
        height: '100vh',
        width: '100vw',
      }}
    >
      <Header />
      <Chooser />
      {Object.entries(gs.icons).map(([key, icon]) => (
        <DraggableIcon icon={icon} id={key} key={key} />
      ))}
    </div>
  );
};
