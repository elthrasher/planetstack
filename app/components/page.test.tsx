import { render } from '@testing-library/react';
import { Page } from './page';
import React from 'react';
import { WSContext } from '../providers/WSProvider';

describe('main page', () => {
  test('render the page', () => {
    const { container } = render(
      <WSContext.Provider
        value={{ gs: { bg: 0, icons: { 'abc-123': { img: 0, x: 400, y: 400 } } }, sendMessage: jest.fn() }}
      >
        <Page />
      </WSContext.Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
