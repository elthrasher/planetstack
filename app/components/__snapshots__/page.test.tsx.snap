// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main page render the page 1`] = `
<div>
  <div
    style="background-repeat: no-repeat; background-size: cover; display: block; height: 100vh; width: 100vw;"
  >
    <div
      class="container header"
    >
      <button>
        background
      </button>
      <button>
        add icon
      </button>
      <button>
        instructions
      </button>
    </div>
    <img
      class="thing react-draggable"
      src="assets/RandomBuildings/B01.png"
      style="transform: translate(400px,400px);"
    />
  </div>
</div>
`;
